packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,338 @@
+import type { Peer } from "@libp2p/interface-peer-store";
+import type { IncomingStreamData } from "@libp2p/interface-registrar";
+import type {
+  Callback,
+  ContentTopic,
+  IAsyncIterator,
+  IDecodedMessage,
+  IDecoder,
+  IFilterV2,
+  IProtoMessage,
+  IReceiver,
+  PeerIdStr,
+  PeerSubscription,
+  ProtocolCreateOptions,
+  ProtocolOptions,
+  SubscriptionsLog,
+  Unsubscribe,
+} from "@waku/interfaces";
+import { WakuMessage } from "@waku/proto";
+import { groupByContentTopic, toAsyncIterator } from "@waku/utils";
+import debug from "debug";
+import all from "it-all";
+import * as lp from "it-length-prefixed";
+import { pipe } from "it-pipe";
+
+import { BaseProtocol } from "../../base_protocol.js";
+import { DefaultPubSubTopic } from "../../constants.js";
+
+import {
+  FilterPushRpc,
+  FilterSubscribeResponse,
+  FilterSubscribeRpc,
+} from "./filter_rpc.js";
+
+const log = debug("waku:filter:v2");
+
+const FilterV2Codecs = {
+  SUBSCRIBE: "/vac/waku/filter-subscribe/2.0.0-beta1",
+  PUSH: "/vac/waku/filter-push/2.0.0-beta1",
+};
+
+class Subscription {
+  private readonly peer: Peer;
+  private readonly pubSubTopic: string;
+
+  private newStream: (peer: Peer) => Promise<Stream>;
+
+  constructor(
+    pubSubTopic: string,
+    remotePeer: Peer,
+    newStream: (peer: Peer) => Promise<Stream>
+  ) {
+    this.peer = remotePeer;
+    this.pubSubTopic = pubSubTopic;
+    this.newStream = newStream;
+  }
+
+  async subscribe<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[],
+    callback: Callback<T>
+  ): Promise<void> {
+    const decodersArray = Array.isArray(decoders) ? decoders : [decoders];
+    const contentTopics = Array.from(groupByContentTopic(decodersArray).keys());
+
+    const stream = await this.newStream(this.peer);
+
+    const request = FilterSubscribeRpc.createSubscribeRequest(
+      this.pubSubTopic,
+      contentTopics
+    );
+
+    try {
+      const res = await pipe(
+        [request.encode()],
+        lp.encode,
+        stream,
+        lp.decode,
+        async (source) => await all(source)
+      );
+
+      const { statusCode, requestId, statusDesc } =
+        FilterSubscribeResponse.decode(res[0].slice());
+
+      if (statusCode < 200 || statusCode >= 300) {
+        throw new Error(
+          `Filter subscribe request ${requestId} failed with status code ${statusCode}: ${statusDesc}`
+        );
+      }
+
+      log(
+        "Subscribed to peer ",
+        this.peer.id.toString(),
+        "for content topics",
+        contentTopics
+      );
+    } catch (e) {
+      throw new Error(
+        "Error subscribing to peer: " +
+          this.peer.id.toString() +
+          " for content topics: " +
+          contentTopics +
+          ": " +
+          e
+      );
+    }
+
+    const subscription: PeerSubscription<T> = {
+      callback,
+      decoders: decodersArray,
+      pubSubTopic: this.pubSubTopic,
+    };
+    FilterV2.subscriptionsLog.set(
```

Comment:
This side-effect (changing `FilterV2` from `Subscription`) seems unnecessary, just add the subscription to `subscriptionsLog` when you create it in `createSubscription`.

Reply:
Also, before creating a subscription, you should probably check if there's already a subscription existing for the given pubsubTopic+peerId. Because if you have two `Subscription` objects for the same pubsubtopic+peerId, the remote peer will only see it as one and any change done by one `Subscription` will modify the other.

That, again, should be done in `createSubscription`.